llm/kv_cache.go

```diff
@@ -0,0 +1,69 @@
+package llm
+
+import (
+	"log/slog"
+	"slices"
+)
+
+// Interface for GGML functionality needed by validation
+type GGMLModel interface {
+	KV() KV
+}
+
+// kVCacheQuantization determines the appropriate KV cache quantization type based on
+// model characteristics, GPU capabilities, and environment settings
+func kVCacheQuantization(kvct string, ggml GGMLModel) string {
+	if kvct == "f16" || kvct == "f32" {
+		return kvct
+	}
+
+	// First check if the requested type is valid
+	validKVCacheTypes := []string{"f32", "f16", "q8_0", "q4_0"}
+	if !slices.Contains(validKVCacheTypes, kvct) {
+		slog.Warn("unsupported cache type, defaulting to f16", "type", kvct)
+		return "f16"
+	}
+
+	// Check flash attention support
+	flashAttnEnabled, _ := checkFlashAttentionSupport(ggml)
+	if !flashAttnEnabled && kvct != "f16" && kvct != "f32" {
+		slog.Info("quantized KV cache requires flash attention, defaulting to f16",
+			"requested_type", kvct)
+		return "f16"
+	}
+
+	// Flash attention is enabled and type is valid, use requested type
+	return kvct
+}
+
+// estimateKvCacheSize determines the total memory required for KV cache based on the quantization type
+func estimateKvCacheSize(cacheType string, numCtx uint64, ggml *GGML) uint64 {
```

Comment:
See note below on integrating this closer with the kv calculations in `ggml.GraphSize`